be/src/exec/pipeline/scan/connector_scan_operator.cpp

```diff
@@ -26,17 +26,59 @@ namespace starrocks::pipeline {
 
 // ==================== ConnectorScanOperatorFactory ====================
 
-ConnectorScanOperatorFactory::ConnectorScanOperatorFactory(int32_t id, ScanNode* scan_node, size_t dop,
-                                                           ChunkBufferLimiterPtr buffer_limiter)
+struct ConnectorScanOperatorIOTasksMemLimiter {
+    mutable std::shared_mutex lock;
+
+    size_t dop = 0;
+    int64_t mem_limit = 0;
+    int64_t running_chunk_source_count = 0;
+    int64_t estimated_mem_usage_update_count = 0;
+    double estimated_mem_usage_per_chunk_source = 0;
+
+    int available_chunk_source_count() const {
+        std::shared_lock<std::shared_mutex> L(lock);
+
+        int64_t max_count = mem_limit / estimated_mem_usage_per_chunk_source;
+        int64_t avail_count = (max_count - running_chunk_source_count) / dop;
+        avail_count = std::max<int64_t>(avail_count, 0);
+        if (avail_count == 0 && running_chunk_source_count == 0) {
+            avail_count = 1;
```

Comment:
If return avail_count=0, what will happen?

Reply:
It means we will hit memory limit.